src/asset_util/src/lib.rs

```diff
@@ -0,0 +1,420 @@
+#[derive(Debug, Default, Clone)]
+pub struct CertifiedAssets {
+    pub assets: HashMap<String, (Vec<HeaderField>, Vec<u8>)>,
+    pub certification_v1: RbTree<String, Hash>,
+    pub certification_v2: NestedTree<Vec<u8>, Vec<u8>>,
+}
+
+impl CertifiedAssets {
+    /// Returns the root_hash of the asset certification tree.
+    pub fn root_hash(&self) -> Hash {
+        fork_hash(
+            // NB: Labels added in lexicographic order.
+            &labeled_hash(LABEL_ASSETS_V1, &self.certification_v1.root_hash()),
+            &labeled_hash(LABEL_ASSETS_V2, &self.certification_v2.root_hash()),
+        )
+    }
+
+    pub fn witness_v1(&self, path: &str) -> HashTree {
+        let witness = self.certification_v1.witness(path.as_bytes());
+        fork(
+            labeled(LABEL_ASSETS_V1, witness),
+            pruned(labeled_hash(
+                LABEL_ASSETS_V2,
+                &self.certification_v2.root_hash(),
+            )),
+        )
+    }
+
+    pub fn witness_v2(&self, absolute_path: &str) -> HashTree {
+        assert!(absolute_path.starts_with('/'));
+
+        let mut path: Vec<String> = absolute_path.split('/').map(str::to_string).collect();
+        path.remove(0); // remove leading empty string due to absolute path
+        path.push(EXACT_MATCH_TERMINATOR.to_string());
+        let path_bytes: Vec<Vec<u8>> = path.iter().map(String::as_bytes).map(Vec::from).collect();
+        let witness = self.certification_v2.witness(&path_bytes);
+
+        fork(
+            pruned(labeled_hash(
+                LABEL_ASSETS_V1,
+                &self.certification_v1.root_hash(),
+            )),
+            labeled(LABEL_ASSETS_V2, witness),
+        )
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum ContentEncoding {
+    Identity,
+    GZip,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[allow(clippy::upper_case_acronyms)]
+pub enum ContentType {
+    HTML,
+    JS,
+    JSON,
+    ICO,
+    WEBP,
+    CSS,
+    OCTETSTREAM,
+    PNG,
+    SVG,
+    WOFF2,
+}
+
+impl ContentType {
+    pub fn to_mime_type_string(self) -> String {
+        match self {
+            ContentType::HTML => "text/html".to_string(),
+            ContentType::JS => "text/javascript".to_string(),
+            ContentType::JSON => "application/json".to_string(),
+            ContentType::CSS => "text/css".to_string(),
+            ContentType::ICO => "image/vnd.microsoft.icon".to_string(),
+            ContentType::WEBP => "image/webp".to_string(),
+            ContentType::OCTETSTREAM => "application/octet-stream".to_string(),
+            ContentType::PNG => "image/png".to_string(),
+            ContentType::SVG => "image/svg+xml".to_string(),
+            ContentType::WOFF2 => "application/font-woff2".to_string(),
+        }
+    }
+}
+
+// The <script> tag that loads the 'index.js'
+const JS_SETUP_SCRIPT: &str = "let s = document.createElement('script');s.type = 'module';s.src = '/index.js';document.head.appendChild(s);";
```

Comment:
does it really make sense to make this part of the lib? I feel like this is very specific